# DBSCAN
Implementation of an ML algorithm, DBSCAN, on metabolome abundance data in treatment naive GBM patients.

Description of the data:
Proteogenomic and metabolomic characterization of human glioblastoma. Whole genome or whole exome sequencing of 99 samples. Generated by CPTAC. 

As seen in "Proteogenomic and metabolomic characterization of human glioblastoma. Cancer cell" by Wang LB, Karpova A, Gritsenko MA, et al. (doi:https://doi.org/10.1016/j.ccell.2021.01.006),
Metabolite identifications and data processing were conducted as previously detailed (Webb-Robertson et al., 2014). GC-MS raw data files were processed using Metabolite Detector software v2.0.6 beta (Hiller et al., 2009). Retention indices (RI) of detected metabolites were calculated based on the analysis of the FAMEs mixture, followed by their chromatographic alignment across all analyses after deconvolution. Metabolites were identified by matching experimental spectra to an augmented version of the Agilent Fiehn Metabolomics Retention Time Locked (RTL) Library (Kind et al., 2009), containing spectra and validated retention indices. All metabolite identifications were manually validated. The NIST 08 GC-MS library was also used to cross validate the spectral matching scores obtained using the Agilent library and to provide identifications for metabolites that were initially unidentified. The three most abundant fragment ions in the spectra of each identified metabolite were automatically determined by Metabolite Detector, and their summed abundances were integrated across the GC elution profile. A matrix of identified metabolites, unidentified metabolite features, and their corresponding abundances for each sample in the batch were exported for statistics. 

# EXECUTIVE SUMMARY
The project aimed to perform metabolomic characterisation and analysis of Glioblastoma. Glioblastoma is a highly aggressive brain tumor with complex metabolic dysregulation, making it crucial to understand the underlying metabolomic patterns to gain prospective therapeutic insights. 

High dimensional metabolic data was preprocessed, and clustered using the DBSCAN algorithm to describe metabolic abnormalities. Through principal component analysis (PCA) and differential expression analysis, we aimed to uncover subtypes and compare the data obtained from 99 GBM patients.

Cancer cells often exhibit altered metabolic pathways to support rapid growth and proliferation. The analysis was performed on various attributes:
- comparison of lactic acid, pyruvate, glucose levels
- comparison of homocysteine and creatinine levels 
- comparison of galactitol and glucose levels 
- clustering based on age and BMI
 

The project utilized Python libraries such as pandas, NumPy, and matplotlib to preprocess and visualize metabolomic data. The knee point concept was used to determine the optimal value of the epsilon parameter. 

Visualizations such as heatmaps and pathway maps elucidated disrupted metabolic pathways in GBM, providing critical insights into potential targets and biomarkers. This project highlights Python as a powerful tool for bioinformatics and metabolomics research, enabling comprehensive exploration and comparison of metabolite expression patterns in Glioblastoma.


# INTRODUCTION
Density-Based Spatial Clustering of Applications with Noise (DBSCAN) is a base algorithm for density-based clustering. It is a popular unsupervised learning method used for model construction and machine learning algorithms. It is a clustering method utilized for separating high-density clusters from low-density clusters with data containing outliers and noise. It is very useful in defining arbitrarily shaped clusters. It classifies points as core, border, or noise based on two parameters: Eps (the radius defining a point's neighborhood) MinPts (the minimum number of points required within this radius for a point to be considered a core point)DBSCAN is a very powerful and versatile clustering technique and has a wide range of applications including image processing, geospatial data analysis ,data mining , astronomy and anomaly detection.



# OBJECTIVE
-Implement DBSCAN with optimized parameters using the knee point method for clustering
-Analyze GBM metabolomic profiles and identify subclasses based on the patterns visualized. 
-Implement DBSCAN with optimized parameters using the knee point method for clustering
-Analyze GBM metabolomic profiles and identify subclasses based on the patterns visualized. 
-The analysis was performed on various attributes:
. comparison of lactic acid, pyruvate, glucose levels
. comparison of homocysteine and creatinine levels 
. comparison of galactitol and glucose levels 
. clustering based on age and BMI


# DETAILED DESIGN ARCHITECTURE
## Proposed System Architecture
The proposed system architecture is designed to detect unique metabolic subgroups in Glioblastoma Multiforme (GBM) patients (treatment-naive) by grouping metabolite abundance profiles using the DBSCAN algorithm. The system is intended to take high-dimensional metabolomic data, preprocess it, and perform clustering to identify underlying metabolic abnormalities and possible subgroups in GBM. The system architecture's main components include data collection, preprocessing, clustering algorithm application, visualization and analysis.
![Screenshot 2024-06-14 225247](https://github.com/aditirk1/DBSCAN/assets/132145522/f62f8f91-255a-471b-9b48-bae51725e8d9)


## Design Architecture

Python is a multi-paradigm language that supports various programming approaches, including object-oriented, procedural, and functional styles. It is an interpreted language, meaning that the code is executed directly without being compiled into machine code first. This interpreted nature makes Python highly interactive, allowing developers to test and experiment with code in real-time. In this code, the architecture follows a modular approach.
The Data Input Module encapsulates functionality to handle input data files, ensuring compatibility across various formats and detecting potential errors or inconsistencies during file validation. It transforms input data into a structured format, typically a pandas DataFrame, to facilitate seamless processing. The Exploratory Data Analysis Module provides tools for computing descriptive statistics, generating correlation matrices, and visualizing data through heatmaps and scatter plots. This allows in uncovering underlying patterns and potential clusters within the metabolomic data. The Data Normalization and Scaling Module is used in preprocessing by standardizing feature scales using techniques like Min-Max scaling or standardization, ensuring uniformity across all features. The Dimensionality Reduction Module employs algorithms such as Principal Component Analysis (PCA) to reduce the dimensionality of high-dimensional data while preserving variance, thus enhancing visualization and facilitating effective clustering. The Clustering Module implements the DBSCAN algorithm for identifying clusters based on optimized parameters, determined using techniques like the KneeLocator algorithm for epsilon selection. Lastly, the Visualization enables the visualization of clustering results, creating scatter plots to highlight different clusters and distinguish between noise and cluster points, thus providing a visual aid for identifying subgroups.

## Methodology

1. __Data Acquisition and Preprocessing__

The first step was to load the metabolomic data from a CSV file using pandas for efficient data handling and manipulation. Data integrity and completeness must be maintained by handling missing values through mean imputation. Finally, the dataset was transposed to align metabolite features as columns and patient samples as rows, optimizing the data structure for subsequent analysis.

2. __Merging necessary files__

Clinical patient data consisting of various categorical and numeric data (which were the features used in the analysis) was merged with metabolome abundance data based on the column “Patient_ID”. This ensured that both data were retained, with missing data being depicted as “NaN”. This step was performed for other analysis apart from the PCA derived data.

3. __Exploratory Data Analysis (EDA)__

Following the loading of required data, descriptive statistics using pandas to summarize the distribution of metabolite abundances across samples, including metrics such as mean, standard deviation, and quartiles, was performed. Utilized seaborn and matplotlib for generating correlation matrices and visualizing interrelationships among metabolites using heatmaps and scatter plots. These visualizations can determine potential metabolic patterns and outliers.

4. __Data Normalization and Scaling__
   
Normalization of metabolite abundances using MinMaxScaler from scikit-learn was performed to standardize data across a uniform range, preparing it for downstream analysis. This step ensures that all features contribute equally to algorithms and prevents biases due to varying scales.

5. __Dimensionality Reduction__
    
Implementation of Principal Component Analysis (PCA) from scikit-learn was done to reduce the dimensionality of the dataset while retaining essential variance. For the rest of the analysis, PCA was not performed as there were two dimensions only, when comparing two features in the data. Project the data into a lower-dimensional space using PCA, facilitating efficient visualization and clustering. This simplifies the interpretation of complex metabolomic profiles.

6. __Clustering Analysis__
    
Utilized the Density-Based Spatial Clustering of Applications with Noise (DBSCAN) algorithm from scikit-learn for clustering analysis. Optimized parameters such as epsilon and minimum samples using techniques like KneeLocator for enhanced clustering accuracy. Distinct metabolic subgroups within the data based on clustering results were identified, distinguishing noise from cluster points for comprehensive analysis.

7. __Visualization and Interpretation__
    
Visualized clustering outcomes using matplotlib and seaborn to create scatter plots and color-coded clusters. This visualization approach helps in effectively interpreting metabolic subgroups and highlighting important differences between clusters. 


# RESULTS
GBM cells, like most cancer cells, exhibit the Warburg effect, where they produce energy through glycolysis followed by lactic acid fermentation even in the presence of Oxygen. This allows us to distinguish between cancer patients and healthy people, by analyzing their metabolic profiles. 
The chosen attributes are crucial in energy producing metabolic pathways. Comparison of lactic acid, pyruvate levels between the patients gives an understanding about shift towards glycolysis. 
Homocysteine is an intermediate in the methionine metabolism and is associated with oxidative stress. Creatinine level perturbations can give an insight into the overall metabolic health. Galactitol accumulation can be indicative of a dysregulation of galactose metabolism.
Age can affect the progression of GBM. Clustering based on age can help in identification of age- specific metabolic patterns.
#future enhancement
The goal of future developments for DBSCAN is to increase its usability, performance, and flexibility in a variety of contexts. One of the main areas of improvement is the development of algorithms for automated and dynamic parameter estimation, which optimizes parameter selection and allows for better handling of datasets with different densities. Enhancing parallel processing methods for multi-core and GPU-based systems and integrating DBSCAN into distributed computing frameworks such as Apache Spark are two ways to boost scalability.The future of DBSCAN involves making it more user-friendly, faster, and versatile across different applications. Automated parameter estimation and improved noise handling, will enhance our ability to analyze and interpret complex biological data.



